Improve stdlib support for option values

dbuenzli opened the pull request:
This PR improves support stdlib support for option values. It adds an `Option` module to the standard library and the `Format.pp_print_option` function.

I have reviewed the usual standard library extension libraries. Those tend to define quite a few other functions. I didn't find these particularly enlighting from a code reading point of view: the type being simple, pattern matching on the datatype to indicate what you want to do is often clearer than calling a named function whose behaviour you have to remember in your head. So the definitions are kept to an essential (in my eyes) feature set.

dbuenzli committed:
Stdlib: add Option module.

pmetzger commented:
The exact functions to be included (and their names) can be debated, but the usefulness of something like this seems obvious.

Drup commented on stdlib/option.mli:
```diff
@@ -0,0 +1,64 @@
+(** [some v] is [Some v]. *)
+
+val evict : none:'a -> 'a option -> 'a
+(** [evict ~none o] is [none] if [o] is [None] and [v] if [o] is [Some v]. *)
```
On one hand, this name is typical "bunzli", which makes it very amusing. On the other hand, it's completely unhelpful. `may` is sometimes use for this, but I don't find it particularly clear either. I quite like Container's convention which uses the `_or` prefix for operation that define a default value, which gives you `get_or` and `map_or`.

rgrinberg replied:
The signature in cute should also be considered:

```
value : 'a option -> default:'a -> 'a
```

In my opinion, both core and containers chose better names here.

dbuenzli replied:
I did use `Option.value` in the past but eventually dismissed it: I don't find it compelling since `option`s are also values themselves. 

Regarding the label the advantage of using `~none` rather than `~default` is that you can consistently use it everywhere to denote what happens in this case. E.g. `Option.to_result`, `Format.pp_print_option`, or whatever may be added in the future that needs a special argument for `None`.

pmetzger replied:
`extract` perhaps? `evict` reads very weirdly to a native English speaker, I would assume it would destroy something rather than extracting the wrapped value.

kennetpostigo replied:
`evict` doesn't make much sense to me either

pmetzger replied:
I've also seen `project` used for this (in its sense as a verb rather than as a noun, that is, the act of projection) but I think it's not an obvious word to people who haven't heard terms like the first and second projections from pairs...

dbuenzli replied:
What is so strange about `Option.evict` ? It means you are [evicting](https://www.merriam-webster.com/dictionary/evict) or expelling the option; but then I'm not a native english speaker. 

As for myself I don't find `Option.extract` to be meaningful you are not extracting the option you are extracting a value out of the option.

In any case please provide full signatures with doc strings for alternate suggestions. For now the only other good candidate to my eyes would be @drup's suggestion albeit with a `none` label rather than `default` for the reasons mentioned above, i.e. :
```ocaml
val get_or : none:'a -> 'a t -> 'a
(** [get_or ~none o] is [none] if [o] is [None] and [v] if [o] is [Some v].
```

dbuenzli replied:
Here are a few alternatives names all with the signature given above to be read in an example context. I might post an @mseri style poll at some point: 

```ocaml
let v = Option.evict ~none:x (Map.find_opt k m)
let v = Option.get_or ~none:x (Map.find_opt k m)
let v = Option.remove ~none:x (Map.find_opt k m) 
let v = Option.expel ~none:x (Map.find_opt k m)
let v = Option.elim ~none:x (Map.find_opt k m)
```

I'm quite tempted by simply pushing `Option.remove` though.

bluddy replied:
My vote would be for 1. `Option.unwrap` (currently not in the list) 2. `Option.get_or`
The keyword argument though should be `default` rather than `none` IMO, as the former is much clearer.

yallop replied:
The Haskell equivalent is called [`fromMaybe`](https://hackage.haskell.org/package/base-4.11.1.0/docs/Data-Maybe.html#v:fromMaybe), which suggests  `from_option` or `of_option`.

dbuenzli replied:
@yallop doesn't `Option.{of,from}_option` read very weirdly to you ? It seems at odds with the OCaml `M.of` conventions.

@bluddy I explained why the argument is `none` (note that this pattern also works for e.g. `result`, see the various labellings in the API of [`rresult`](http://erratique.ch/software/rresult/doc/Rresult.R.html)) do you suggest: 

1. To be inconsistent and use `~default` only in that function and `~none` in the others ?
2. To use `~default` in all of these other functions; I doubt that it is a great idea.

If `none` is mentioned in the name though we get could then label with `use`, something like:

```ocaml
let v = Option.on_none ~use:x (Map.find_opt k m) 
```

yallop replied:
@dbuenzli: it reads okay to me, although I can see how it seems at odds with the other `of` functions.

One more possibility (in case there aren't enough under consideration already):  the SML basis uses [`getOPT`](http://sml-family.org/Basis/option.html#OPTION:SIG:SPEC); we could use `Option.get` here, and drop the function that raises `Invalid_argument`.

dbuenzli replied:
Actually those two (raising and non-raising) are those I miss the most...

pmetzger replied:
Having a raising and a non-raising version would be good, yes.

As for this:

> What is so strange about `Option.evict`?

To a native speaker, eviction has completely incorrect connotations. One evicts a person who is late on their rent, for example, and the connotation is not of extracting something to examine it but rather of banishment. The word really isn't correct. One is not evicting the value (which sounds like what you would do when removing an entry from a hash table), one is fetching the value. Words like fetch, get, peek, look, examine, extract, etc. come to mind. Evict really means "banish", and you are not banishing the value, you're getting it.

Drup reviewed (commented):
Thanks for volunteering for doing this one, @dbuenzli , it is highly appreciated. :)

I think the subset of functions you choose is perfectly fine. I would only like two additional ones, `iter` and `to_seq`.

Armael commented:
I agree with @dbuenzli that in many cases it is enough to just use pattern matching, and most of the time I do not feel the need for many option-related combinators.
There is however one infix operator defined in Batteries which I am quite fond of: `(|?) : 'a option -> 'a -> 'a` -- I find it nice and natural to use in `|>` pipelines. I wonder what people think about it? Personally that's maybe the only option combinator I really want to use :).

dbuenzli commented:
> The exact functions to be included (and their names) can be debated, but the usefulness of something like this seems obvious.

@pmetzger concrete contributions to the discussion are welcome but comments like this bring nothing but noise in my opinion. May I ask you to refrain doing them ? `github` reactions are perfect for this.

@drup Added `Option.to_seq` and `Option.iter`. Regarding `evict` I think it's a good name, for now I'll let the bike-shedding continue on that one. If you have a strong opinion please make a precise signature proposal so that if that needs to happen we can finally list a few contenders to choose from.

@armael Except for constructing some values (e.g. `{Map,Set}.t`), I'm personally not very fond of the point-free pipelining style (especially some people tend to use totally gratuitously, I saw some `let y = x |> f` in the wild, not kidding). Unless there's a strong desire I'll keep your request out for now.

Armael commented:
Fair enough. To be clear, I'm not particularly advocating for the pipelining style -- it's just that *when* I use it (which still happens sometimes), I find it convenient to have `|?` to deal with options.

xavierleroy commented:
Adding a back reference to #1920.

xavierleroy commented on stdlib/option.ml:
```diff
@@ -0,0 +1,39 @@
+let get = function Some v -> v | None -> invalid_arg "option is None"
+let map f o = match o with None -> None | Some v -> Some (f v)
+let flat_map f o = match o with None -> None | Some v -> f v
+let flatten = function Some (Some _ as o) -> o | _ -> None
```
`flat_map` is also known as monadic bind.  Could we make sure this Option module will work with whatever PPX implementing monads we have today?  And perhaps with a future, long-overdue "monadic let" syntax?

dbuenzli replied:
I can certainly add a `bind` and a `return` function if that's desired, but then which signature should be supported ? 

A quick search in `opam` suggest there are already at least four `ppx`s to this: `ocaml-monadic`, `omonad`,  `ppx_monadic` and `ppx_let` and as far as I can tell the interface needed by the latter is incompatible with the interface of the others.

Couldn't we simply add these name aliases in a `Monad` submodule once this future, long-overdue "monadic let" syntax lands and/or a proper `Monad.S` signature has been defined in the stdlib ?

lpw25 replied:
Can I suggest just removing `flat_map` from this PR? The name really comes from the equivalent function on lists, but we don't actually have that function in the standard library. By removing it we can avoid getting into issues about how people want the standard library to deal with monads and other computation abstractions on a PR that is about bringing support for options into line with support for other built-in types.

dbuenzli replied:
@lpw25 I don't understand your concern and request:

1.  `flat_map` as defined in this PR would in any case not be a good candidate for a monadic interface due to the order in which I wrote it (`bind` has the argument swapped). 
2. I wouldn't mind if `List.flat_map` was actually added to the stdlib. 
3. I don't see why adding these functions poses a problem and how not adding these useful functions "avoid[s] getting into issues about how people want the standard library to deal with monads and other computation abstractions ".

I mean at that rate why don't you tell me to remove `Option.flatten`, this is just monadic `join` in disguise.

lpw25 replied:
> I wouldn't mind if List.flat_map was actually added to the stdlib.

I have no particular opinion on whether `List.flat_map` should be added or not -- my point is just that I would rather see `List.flat_map` and `Option.flat_map` added together in a different PR -- where discussions about how they should named and the order of their arguments can be discussed. Essentially, I find it very strange to have `Option.flat_map` without `List.flat_map` because the name `flat_map` derives from the list version of the function.

> I mean at that rate why don't you tell me to remove Option.flatten

Because we already have `List.flatten` so adding `Option.flatten` is consistent.

dbuenzli replied:
Not really convinced but I'm a pleasing machine this evening. Function removed from the PR.

dbuenzli replied:
(and for interested persons I have no plan to carry the work for `{List,Option}.flat_map` myself)

gasche commented:
@Armael: one problem with `|? : 'a option -> 'a -> 'a` is that it computes the default value (the second argument) even when it is not needed, which makes it a potential source of performance bad-surprises.

hcarty commented on stdlib/format.mli:
```diff
@@ -949,6 +949,15 @@ val pp_print_text : formatter -> string -> unit
+  ?none:(formatter -> unit -> unit) ->
+  (formatter -> 'a -> unit) -> (formatter -> 'a option -> unit)
+(** [pp_print_option ?none pp_v ppf o] prints [o] on [ppf]
+    using [pp_v] if [o] is [Some v] and [none] if it is [None] (defaults
```
Maybe `[none] prints nothing by default`?

dbuenzli replied:
Thanks. Tweaked.

pmetzger commented:
The Menhir option.ml/mli also has `fold` (and `hash`). See: https://gitlab.inria.fr/fpottier/menhir/blob/master/src/option.mli

dbuenzli commented:
Now that I added `Option.iter` I guess I can add `Option.fold` though I wonder if people actually use that stuff. I used Menhir's signature as it correspond to the order in `{Map,Set}.fold`; it is however not the order of `CCOpt.fold`.

I also added `Option.hash`. To avoid making `Option` dependent on the `Hashtbl` module I simply hash `None` to `0`, a hash expert should tell me if that's ok or if I should rather reorder the dependencies and use `Hashtbl.hash` on `None`.

yallop commented on stdlib/option.mli:
```diff
@@ -0,0 +1,73 @@
+(** [map f o] is [None] if [o] is [None] and [Some (f v)] is [o] is [Some v]. *)
+
+val flatten : 'a option option -> 'a option
+(** [flatten oo] is [Some v] is [oo] is [Some (Some v)] and [None]
```
typo: `is [oo]` ↝ `if [oo]`

yallop commented on stdlib/option.mli:
```diff
@@ -0,0 +1,73 @@
+(** [to_result ~none o] is [Ok v] if [o] is [Some v] and [none] if *)
+
+val of_result : ('a, _) result -> 'a option
+(** [of_result r] is [Some v] is [r] is [Ok v] and [None] otherwise. *)
```
typo: `is [r]` ↝ `if [r]`

garrigue commented:
+1 for `Option.unwrap`, since it reads nicely as a qualified identifier.

For what it is worth, lablgtk internally uses the following functions:
```ocaml
val may : f:('a -> 'b) -> 'a option -> unit
val may_map : f:('a -> 'b) -> 'a option -> 'b option
val default : 'a -> opt:'a option -> 'a
val may_default : ('a -> 'b) -> 'a -> opt:'b option -> 'b
````

yallop commented on stdlib/option.mli:
```diff
@@ -0,0 +1,73 @@
+(** {1:convert Converting} *)
+
+val to_result : none:'e -> 'a option -> ('a, 'e) result
+(** [to_result ~none o] is [Ok v] if [o] is [Some v] and [none] if *)
```
typo: `[none] if *)` ↝`[none] otherwise. *)`

alainfrisch commented on stdlib/option.mli:
```diff
@@ -0,0 +1,73 @@
+(** {1:convert Converting} *)
+
+val to_result : none:'e -> 'a option -> ('a, 'e) result
+(** [to_result ~none o] is [Ok v] if [o] is [Some v] and [none] otherwise. *)
```
... and `[Error none]` otherwise...

alainfrisch commented on stdlib/option.mli:
```diff
@@ -0,0 +1,73 @@
+(** [of_result r] is [Some v] if [r] is [Ok v] and [None] otherwise. *)
+
+val to_seq : 'a option -> 'a Seq.t
+(** [to_seq o] is [o] as a sequence. *)
```
Perhaps this is obvious enough for most participants to this discussion, but making it explicit that `Some` (resp. `None`) produces a sequence of length 1 (resp 0) might be useful.

alainfrisch commented:
Independently of the naming for the function and the argument, what about using a single function for raising and non raising variants:
````ocaml
val get: ?default:'a -> 'a option -> 'a
````
One could argue that keeping them separate makes it easier to grep for the raising variant, or mark its type as such (when typed effects will be there).

~~Of course, one should also have a non-raising, option returning variant to be coherent with the rest of the stdlib~~

alainfrisch commented on stdlib/option.mli:
```diff
@@ -0,0 +1,75 @@
+
+(** {1:preds Predicates and comparisons} *)
+
+val is_none : 'a option -> bool
```
It's completely minor, but to be coherent with `of_result`, you should either use `_` for all ignored type variables (`val is_none : _ option -> bool`) or none of them (`val of_result : ('a, 'b) result -> 'a option`).  I believe using `_` might be slightly confusing for beginner, so I'd have a preference for introducing (useless) variable names.

dbuenzli replied:
Done (named it so that comment won't outdate itself).

alainfrisch commented on stdlib/option.mli:
```diff
@@ -0,0 +1,75 @@
+
+val compare : ('a -> 'a -> int) -> 'a option -> 'a option -> int
+(** [compare cmp o0 o1] is a total order on options using [cmp] to compare
+    values wrapped by [Some _]. *)
```
I'd suggest documenting the respective ordering of Some vs None.

dbuenzli commented:
@alainfrisch that's an interesting `get` proposal but I worry this might be slightly more error prone to write and harder to keep in the head when reading or looking for a bug.

alainfrisch commented on stdlib/option.mli:
```diff
@@ -0,0 +1,75 @@
+
+(** {1:options Options} *)
+
+type 'a t = 'a option
```
What about re-exporting constructors here, which would allow explicit qualification `Option.Some` (also in patterns)?

alainfrisch commented on stdlib/option.mli:
```diff
@@ -0,0 +1,75 @@
+(** The type for option values. Either [Some v] or [None]. *)
+
+val some : 'a -> 'a option
+(** [some v] is [Some v]. *)
```
I know this is a bit useless, but having `Option.some` but not `Option.none` feels weird.

dbuenzli commented:
The PR has been commented by a few dev team members and I believe most of the problems mentioned have been taken into account. The last one that remains is `Option.evict` which has to be evicted it seems. 

I trimmed down the many proposals to the following five candidates, please vote for the one(s) you'd like to read for the rest of your life or the last item if you are not happy about the selection. In casting your choice please consider the following conflicting design concerns:

1. Relationship to `Option.get` which raises on `None`.
2. Relationship to similar functionality in a potential `Result` module.
3. Consistent `none` labelling in other functions (`Option.to_result`,  `Format.pp_print_option`, future additions) and labelling scheme in a potential `Result` module.

```ocaml
👍 let v = Option.unwrap ~default:x (Map.find_opt k m)
😄 let v = Option.remove ~none:x (Map.find_opt k m)
🎉 let v = Option.get_or ~use:x (Map.find_opt k m)
❤️ let v = Option.if_none ~use:x (Map.find_opt k m)
😕 let v = Option.value ~default:x (Map.find_opt k m)
👎 let v = failwith "Try harder."
```

xavierleroy commented:
As names of pure functions (without effects) I like nouns better than verbs.  `value` was already proposed (I think).  There is also `contents`:
```
let v = Option.value ~default:x (Map.find_opt k m)
let v = Option.contents ~default:x (Map.find_opt k m)
```

mshinwell commented:
I must say, I find Core's `value ~default` to be clearer than any of the four proposed above (and no, I didn't choose its name).  @dbuenzli 's arguments above about why it is bad don't seem compelling to me.  (In particular, you could argue that the `default` label is better than `none` since it could apply to similar types which don't happen to have the nullary case called `None`...)

The "`if_none ~use`" one is kind of nice but falls down if you put the labelled argument second, as some people will do.

@xavierleroy has also just suggested `contents`, which seems good.

dbuenzli commented:
On popular request I have added `Option.value ~default` to the poll.

trefis commented:
![I'm confused](https://i.imgur.com/wXTnQg8.png)

I'm really confused.

dbuenzli commented:
@trefis No citizen left behind. Use this image to orient yourself:
<img width="685" alt="poll" src="https://user-images.githubusercontent.com/485596/43392668-a81c528a-93f4-11e8-8396-530e7ae11ba0.png">

mseri commented:
i have missed this module some much!!! 

In my votes, I don’t really mind if ~default is replaced by ~none

rizo commented on stdlib/option.mli:
```diff
@@ -0,0 +1,78 @@
+
+(** {1:convert Converting} *)
+
+val to_result : none:'e -> 'a option -> ('a, 'e) result
```
I think it's more important to focus on the _outcome_ in this case, which is the fact that the `None` case is the error value. Thus my suggestion is to rename the labeled argument as in:

```ocaml
val to_result : error:'e -> 'a option -> ('a, 'e) result
```

**Examples**:

```ocaml
assert (Option.to_result (Some 42) ~error:"Oh no!" = Ok 42);
assert (Option.to_result None ~error:"Oh no!" = Error "Oh no!");
```

dbuenzli replied:
I'm not sure I agree with this, I think using `~none` make it more clear that this is what happens on `None`.

trefis replied:
I agree with @dbuenzli on that one.

pmetzger replied:
I found `none` slightly confusing because of the declaration of the `none` function a few lines before, but not especially so. I could go either way. That said, something like `if_none` or `on_error` or what have you strikes me as better still. That said, this is such a minor issue you should take my comment with a kilogram of salt.

rizo commented:
That looks great, @dbuenzli! I have been thinking about proposing a similar module for a while. Here're some of my suggestions regarding this.

### Alternatives to `evict`

I also find the `evict` verb very confusing. My initial interpretation was that it evicts the value wrapped by the option somehow.

My another concern, mentioned briefly by @gasche, is that the default value to replace `None` is always evaluated which is often _not_ desirable.

Having that in mind, here're two complementary functions I propose:

```ocaml
let v = Option.get_or x (Map.find_opt k m)
let v = Option.get_or_else (fun () -> x) (Map.find_opt k m)
```

The "else" indicates that the argument has to be a thunk and will not be evaluated unless needed.

Another (more ambitious) alternative is to use `Option.or_else (fun () -> x)` and reuse the `or` operator for the direct value case:

```ocaml
print_endline (Some "some value" |> Option.or_else input_line);;
print_endline (Some "some value" or "nothing");;
```

But I understand that this might be a bit too much! :)


### Discourage Unsafe `get`

Although it's a common pattern in the Stdlib to rely on exceptions, I think that a short and nice name as `get` will further promote users to rely on exceptions. My suggestion is to rename it to something like `Option.force`, `Option.unwrap` or even `Option.unsafe_get`.


### New Functions

I think the following functions could be added for convenience and consistency. I find `catch` particularly useful because the Stdlib relies extensively on exceptions.

```ocaml
val if_some : ('a ‑> unit) ‑> 'a option ‑> unit
(** [if_some f o] is [f v] if [o] is [Some v] and [()] otherwise. *)

val if_none : (unit ‑> unit) ‑> 'a option ‑> unit
(** [if_none f o] is [f ()] if [o] is [None] and [()] otherwise. *)

val catch : (unit ‑> 'a) ‑> 'a option
(** [catch f] is the value produced by [f ()] wrapped in by [Some] and
    [None] if [f] raises an exception. *)

val to_bool : 'a option ‑> bool
(** [to_bool o] is [true] if [o] is [Some _] and [false] otherwise. *)

val to_list : 'a option -> 'a list
(** [to_list o] is [o] as a list. [None] is the empty list and
    [Some v] is the singleton list containing [v]. *)
```

rizo commented:
I remembered another common use case: when extracting the value it's tempting to use the unsafe method that raises an exception. For that case an informative error message could be added:

```ocaml
Map.find_opt k m |> Option.get_or_fail (Fmt.strf "Key %S not found" k)
```

dbuenzli commented:
> My another concern, mentioned briefly by @gasche, is that the default value to replace None is always evaluated which is often not desirable.

Yes there are a few functions for this in extended libraries but I don't find the names very good: containers has `CCOpt.get_lazy`,  batteries `BatOption.default_delayed`. 

I think that in that case it's better to pattern match and do whatever you want to compute in the `None` branch. This will also avoid closure allocation.

> Although it's a common pattern in the Stdlib to rely on exceptions, I think that a short and nice name as get will further promote users to rely on exceptions.

No this raises an `Invalid_argument` exception. `Invalid_argument` exceptions are for programming errors. `Option.get` is only to be used when you have an invariant that tells you the value is `Some _`; see also `{String,Bytes,Array}.get`.

> val if_some : ('a ‑> unit) ‑> 'a option ‑> unit
> (** [if_some f o] is [f v] if [o] is [Some v] and [()] otherwise. *)
> 
> val if_none : (unit ‑> unit) ‑> 'a option ‑> unit
> (** [if_none f o] is [f ()] if [o] is [None] and [()] otherwise. *)

The first one is `Option.iter`, the second you can use `is_none` and/or pattern match. 

> val catch : (unit ‑> 'a) ‑> 'a option
> (** [catch f] is the value produced by [f ()] wrapped in by [Some] and
>     [None] if [f] raises an exception. *)

Out of question. This implies a catch all exception handler which is dangerous in practice (asynchronous exceptions). And nowadays that we have exception matching the problem of exceptions in the stdlib is much less acute than it used to be.

> val to_bool : 'a option ‑> bool
> (** [to_bool o] is [true] if [o] is [Some _] and [false] otherwise. *)

I don't really see why `None` should necessarily be equated with [false]. Using `Option.is_{some,none}` is much clearer.

> val to_list : 'a option -> 'a list
> (** [to_list o] is [o] as a list. [None] is the empty list and
>     [Some v] is the singleton list containing [v]. *)

Maybe this one could be added but why shouldn't this be `val to_list : 'a option -> 'a option list` ? From a code reading perspective I think these kind of mapping ([None] is ([])) are better understood by a concrete pattern match.

dbuenzli commented:
(If there is strong support for `to_list` I will add it)

rizo commented:
I find `to_list` more useful than `to_seq`.

rizo commented:
> No this raises an Invalid_argument exception. Invalid_argument exceptions are for programming errors. Option.get is only to be used when you have an invariant that tells you the value is Some _; see also {String,Bytes,Array}.get.

I understand this. My point is that the assumption that _"`Option.get` is only to be used when you have an invariant"_ is not enforced in any way at all, which in practice does lead to programming errors. This applies to `{String,Bytes,Array}.get`, `Map.find` as well. Of course, it's not realistic to change those, what I'm asking is to sacrifice consistency with an old habit for safety, in this case.

lpw25 commented:
> This applies to {String,Bytes,Array}.get, Map.find as well

You're conflating two different things here. `Map.find` is very different from `Option.get`. `Map.find` requires an exception handler as part of its usual operation -- this is what encourages error prone code. `Option.get` should never have a related exception handler, which is clear from its use of `Invalid_argument` -- no one should ever match `Invalid_argument` and doing so is a very obvious code smell -- which makes `Option.get` much less problematic.

Drup commented:
> no one should ever match Invalid_argument and doing so is a very obvious code smell -- which makes Option.get much less problematic.

You are correct .... if you assume a familiarity with OCaml idioms. However, if we provide both `is_some` and `get`, I'm predicting we are going to see many code patterns that look like this in the wild:

```
if Option.is_some o then
  let v = Option.get o in
  ...
else
  ...
```

This was very obvious in the Rust community, in particular, were many people coming from more imperative languages would write code like that. Of course, we shouldn't always design our API in term of how beginners are going to misuse them, but it is something to keep in mind for this kind of designs. Especially in these times of opening to a broader audience more familiar with Javascript!

rizo commented:
@lpw25 You're right, thanks for clarification. But I also think this distinction makes almost no difference in practice. Both cases rely on conventions and good practices to avoid exceptions.

I'm suggesting to avoid this problem by giving `get` a less attractive name and encourage pattern-matching instead. I don't think that an experienced OCaml developer would use `get` frequently, but beginners certainly will, as @Drup noted.

dbuenzli commented:
> I find to_list more useful than to_seq.

Right, since it was done for delayed lists there's no reason not to do it for lists. Function added.

lpw25 commented:
> But I also think this distinction makes almost no difference in practice

Why? In my experience it makes a huge difference in practice. 

>  I'm predicting we are going to see many code patterns that look like this in the wild

If they want to write that code then there is nothing wrong with it. It's not idiomatic but it's not broken either, and it's correctness is easy to see.

Really, `Option.get` is the same as `List.hd` and personally I've never found that function's existence with a short name to be a problem.

pqwy commented:
First of all, I'd like to thank @dbuenzli for taking the time to make this PR, and to deal with the responses. Adding a module for a well-known and widely used type is bound to elicit a measure of enthusiasm, and I am grateful for Daniel's stoic attempts to weather this.

I'd like to add two observations, FWIW:

- At first, I was pretty disappointed that there is no `bind` under any name. However, as @lpw25 points out, we don't really have it for other monadic types either. This means keeping ad-hoc versions of it around, but the matter is probably better dealt with across the whole stdlib in one go -- if ever.

- The current signature contains a number of fairly specific combinators, but omits the actual eliminator for `option`. Not getting into the naming issue, the eliminator would be `bacon: 'a -> ('b -> 'a) -> 'b option -> 'a`. This is the type of the actual pattern match, and is intimately connected to `option`. It also allows expressing a number of existing combinators with several key-strokes:

        let hash = bacon 0
        let fold = bacon (fun x -> x)
        let iter = bacon ()
        let flatten = bacon None (fun x -> x)
        ...

    This function is named [`maybe`](http://hackage.haskell.org/package/base-4.11.1.0/docs/Prelude.html#v:maybe) in Haskell, and its analogue is `List.fold_right`. I usually call it `option`.

I think the eliminator should be added at a minimum. I would also reconsider some of the relatively gratuitous specialized combinators currently in the signature.

rizo commented:
> Why? In my experience it makes a huge difference in practice.

Because in practice, in both cases, people forget 1) to handle exceptions when using operations that rely on them (_i.e._ `Map.find`), and 2) to perform the checks needed to safely use the values (_i.e._ `Option.get`).

For what it's worth, I think `List.hd` should return an option, forcing the users to consider the negative case. I consider Core to be a good example of a library with safety enforced in types, as opposed to conventions (which it does in both presented cases, by the way).

This might seem off-topic and obvious, but I believe that the Stdlib should not be designed only with experienced developers in mind. If a function can fail in an unexpected way it should be obvious from it's name. Thus I think _not_ having `Option.get` at all is better than keeping it with the current name.

dbuenzli commented:
@rizo I really think that a function raising `Invalid_argument` should exist in one way or another, for once we can be consistent, I prefer if we do so.

@pqwy 

I won't comment on `bind` I think we'll quickly get into contentious territory, I'll let someone else pick up that fight.

I think adding the eliminator makes sense, batteries, containers and base have the function but under suboptimal names in my opinion (`map_default`, `map_or`, `value_map`). Haskell seems to consistently use the name of the type itself for theses functions: [`maybe`](http://hackage.haskell.org/package/base-4.11.1.0/docs/Prelude.html#v:maybe), [`either`](http://hackage.haskell.org/package/base-4.11.1.0/docs/Prelude.html#v:either) or `fold`s. Scala uses [`Option.fold`](https://www.scala-lang.org/api/current/scala/Option.html#fold[B](ifEmpty:=%3EB)(f:A=%3EB):B) and [`Either.fold`](https://www.scala-lang.org/api/current/scala/util/Either.html#fold[C](fa:A=%3EC,fb:B=%3EC):C), we already have `List.fold_right` for lists so I changed the dubious `fold` that I added in 812fd6d to be that function.

pmetzger commented on stdlib/option.ml:
```diff
@@ -20,7 +20,7 @@ let get = function Some v -> v | None -> invalid_arg "option is None"
 let flatten = function Some (Some _ as o) -> o | _ -> None
 let iter f = function Some v -> f v | None -> ()
-let fold f o acc = match o with Some v -> f v acc | None -> acc
+let fold none some = function Some v -> some v | None -> none
```
Why the change to remove the accumulator?

dbuenzli replied:
You can easily recover that, see discussion above.

rizo commented:
@dbuenzli I'm not against consistency (on the contrary!), I just think safety is more important. In my opinion `Option.get` leads to unsafe code in (almost) the same way `null` pointers do.

Do you think we could at least have a way to pass an error message if the option is `None`?

In Rust this is done with the [`expect`](https://doc.rust-lang.org/std/option/enum.Option.html#method.expect) function.

Here's an example:

```ocaml
let maybe_item = Map.find_opt k m in
let item = Option.expect "key not found in map" maybe_item
```

dbuenzli commented:
> Do you think we could at least have a way to pass an error message if the option is None?

I think bothering programmers with inventing an error message here it not useful. Remember: this is a failed assertion, what you are asserting is `o != None` and that's what the current error message tells you.

A custom error might, if you are lucky, make it a little bit easier to `grep` for the failed assertion in a code base but what you really want here is a stacktrace which IIRC OCaml now has in good shape in almost all backends. Yes some people compile without `-g`, but different birds told me they no longer do it, even in production, because it makes error tracking too hard for the actual performance return.

Regarding the example you gave I think it would be more useful to add `Map.get` (I think it would also be nicer if a pretty-printer was required from keys so that you can give a really good error message but all that is way beyond this discussion).

ghost commented:
Since we have been using an `Option` module at Jane Street for a long time and we have a large code base, I though it could be useful to the discussion to provide some usage statistics. Here are the most used functions from `Option` in the Jane Street code base (~10 millions lines of code):

```
$ ag -o 'Option\.[a-z_A-Z0-9]+'  --no-number |cut -d: -f2|sort | uniq -c|sort -n
    537 Option.equal
    639 Option.Let_syntax
    669 Option.try_with
    996 Option.some_if
   1129 Option.Optional_syntax
   1381 Option.is_none
   1652 Option.bind
   1671 Option.V1
   2183 Option.is_some
   2308 Option.iter
   2997 Option.value_map
   3449 Option.some
   3533 Option.none
   4171 Option.value_exn   # our Option.get
   6196 Option.Stable
   6836 Option.value
   8248 Option.map
  13505 Option.t
```

from these numbers we can see that:
- `Option.get` is widely used in practice and can be used in safe code (I'm assuming that our code is generally safe)
- `Option.value` would be a useful addition

From personal experience, I can confirm that `Option.value`, `Option.map` and `Option.value_exn` are the three functions that I have used the most, especially `Option.value`.

FTR, this is `Option.value`:

```ocaml
(** [value None ~default] = [default] [value (Some x) ~default] = [x] *)
val value : 'a t -> default:'a -> 'a
```

gasche commented:
`value_exn` is a much better name than `get` in term of forcing people to face the consequence of their choice. If we have it at all (personally, I'm not convinced that it is necessary for a first iteration), having the explicit `_exn` would be important, I think.

rizo commented:
> I think bothering programmers with inventing an error message here it not useful. Remember: this is a failed assertion, what you are asserting is o != None and that's what the current error message tells you.

I disagree, programmers will certainly have the context to write a meaningful error message. They need to do it only once, while the error might be triggered multiple times.

I find assertions without an error message equally annoying. I don't think just stacktraces are a good substitute for error messages – their role is to tell _where_ something failed. Getting a generic `Assert_failure` exception is frustrating because it's not clear at all _what_ failed and _why_.

Anyway, I think my point is clear now and I'll leave the decision to you :)

dbuenzli commented:
Thanks @diml for the numbers. If you want `Option.value` please cast your vote [here](https://github.com/ocaml/ocaml/pull/1940#issuecomment-408798769) is competing with `unwrap` (which I personally find not that great, e.g. considering a `Result` module, a potential `Result.value` would be more meaningful to me than `Result.unwrap`).

> value_exn is a much better name than get in term of forcing people to face the consequence of their choice. If we have it at all (personally, I'm not convinced that it is necessary for a first iteration), having the explicit _exn would be important, I think. 

There are so many functions in the stdlib that raise `Invalid_argument` I don't think it would have been an improvement if all of these functions had that labelling and I never thought these kind of `_exn` suffixes were any kind of improvement. I'll stick behind both the need (see @diml stats) and the name for consistency with the rest of the stdlib.

> I disagree, programmers will certainly have the context to write a meaningful error message.

If you want to have a meaningful message you likely want a lazily constructed string at which point youl'll be better served with matching (see discussion about `Option.value`). My experience is that programmers do not write meaningful error message but rather crappy and inconsistent ones especially if they are gratuitously asked to do so. If you are writing libraries and can provide a better error message then you won't use `get` and will likely want a lazily constructed string see discussion about `Map.get`.

trefis commented:
> My experience is that programmers do not write meaningful error message but rather crappy and inconsistent ones especially if they are gratuitously asked to do so. If you are writing libraries and can provide a better error message then you won't use get and will likely want a lazily constructed string

Agreed.

rizo commented:
@dbuenzli Do you plan on proposing the `Result` module? As you suggested, consistency between the two should be taken into account with regards to naming. I'd be glad to make a separate PR, if you don't have time/energy to manage it.

alainfrisch commented:
+1 for not specifying an "error message".  This would give the impression that it's ok to use the function in contexts where the option can be None (and producing some nice user-facing message), which is not the case (i.e. None is a programming bug, in the same way as an out-of-bound array access)

dbuenzli commented:
> @dbuenzli Do you plan on proposing the Result module?

I do plan to propose the uncontroversial and common parts of the various Result modules out there in a separate PR at some point.

dbuenzli commented:
I have the impression that most of the hard points have been discussed. I have thus rebased the work to two commit. In that process: 

1. I have changed `Option.evict` to `Option.value` with the exact signature @diml suggested (the polls seems to be in favor of it aswell at the time I write this).
2. I removed the `Option.hash` function. I think it is better to handle this precisely at the point where you need it (e.g. if you need the seeding stuff).
3. Removed the `Option.of_result` function (this was the only `of_` among all these `to_` functions and will become clearer soon, see below).
4. Became optimist about the version in which that should become available. 

I will now prepare a `PR` for `Result` so that the consistency with certain names can be cross-checked.

mshinwell commented:
I think it's a mistake to remove `Option.hash`.  Even if there is another claimed solution, what I reckon will happen here is simply more use of the polymorphic hash function, which is bad.

dbuenzli commented:
@mshinwell if I get a definitive ok by people who are more into hashes than I am that these two would be good for `option` and `result` I will add them to the proposal(s): 
```
let hash_option : ('a -> int) -> 'a option -> int = Option.fold 0
let hash_result ('a -> int) -> ('e -> int) -> ('a, 'e) result -> int = Result.fold
```
(or whatever else these people would recommend; I do however worry a bit about potentially not applying any form of seeding to `None` and not distinguishing the cases in `hash_result`).

yminsky commented:
These definitions of hash_option and hash_result seem at least a little suspect to me. I wonder if @braibant  or @aalekseyev have any thoughts on this, since they did a lot of work on hashing as part of the development of ppx_hash.

aalekseyev commented:
I would say the main problem with the proposed hash functions is that they create systematic collisions.
In `ppx_hash` we generally followed the rule of never creating systematic collisions (the only collisions are the ones created by the user or by the mixing primitive).

What I'm talking about is `Some None` colliding with `None` and `Ok x` colliding with `Error x`.
So you end up with very big families of collisions in types like `(unit, unit) result list`.

We do is things like:
```
let hash_result hash_a hash_e x = function
  | Ok a -> mix 0 (hash_a a)
  | Error e -> mix 1 (hash_e e)
```

For some reasonable somewhat-collision-resistant mixing primitive `mix`.

dbuenzli commented:
@mshinwell given words of @aalekseyev (thanks !) I think it's better if we leave the hash functions out until suitable `Hashtlb.mix` primitives are added to the stdlib.

dbuenzli commented:
So I rebased the PR in the process:

1. Since [this](https://github.com/ocaml/ocaml/pull/1956/files#r207033079) and [this](https://github.com/ocaml/ocaml/pull/1956/files#r207033017) comment received no objections I unified the interface with the `Result` [PR](https://github.com/ocaml/ocaml/pull/1956). This means replaced `Option.flatten` by `Option.join` and added `Option.bind`.
2. Added a test suite.

dbuenzli commented on testsuite/tests/lib-option/test.ml:
```diff
@@ -0,0 +1,119 @@
+(* TEST
+*)
+
+module Option = Stdlib__option (* why ? *)
```
@shindere Could you please have a look at this. It seems that the testing framework is not able to compile the this file if I don't add this (same problem in `Result` PR). If you take for example `lib-uchar` this is not needed so I suspect a bootstrap issue but I don't think a bootstrap should be needed if you add a module to the stdlib and it's not used by the compiler itself.

Also it would be nice to report the actual compilation error when you fail to compile the `test.ml` file rather than a simple exit code.

nojb replied:
@dbuenzli I could run this test without issue with a fresh build of your branch; maybe try cleaning your tree ?

nojb replied:
(that is, I could run it without the `module Option = ...` line)

dbuenzli replied:
@nojb how clean ?  somehow with `make clean` doesn't work here.

nojb replied:
`git clean -fdx`

nojb replied:
What is the error given by the compiler ?

dbuenzli replied:
Tried `make distclean`, didn't work but `git clean -fdx` did, thanks ! (in case you care the error was that `Option` module didn't exist)

dbuenzli committed:
Add Format.pp_print_option.

nojb commented:
CI green! The PR seems to have reached a stable point, so I am planning to merge this tomorrow unless someone speaks up, or beats me to it.

nojb commented:
@dbuenzli Could you update `Changes` ?

dbuenzli commented:
Done. For credits see [this comment](https://github.com/ocaml/ocaml/pull/1956#issuecomment-411332169).

dbuenzli commented:
I'll rebase that to solve the conflicts now that https://github.com/ocaml/ocaml/pull/1956 is in.

dbuenzli commented:
Done.
